This is a branch for use with a migrated celo datadir

piersy opened the pull request:
This contains all changes required to get op-geth to start up and be able to process all the custom datatypes that existed in celo-blockchain during syncing. This doesn't contain any changes to support access over the RPC.

Signing via a types.Signer and transaction JSON Marshaling (which I believe is only used for tx signing operations), is supported only for CeloDynamicTxV2s (type 123). This code was pre-existing in the celo4 branch, we could further remove this code because it is not required for sync but it seems simpler to leave it as is.

Note we will need to implement types.Signers for legacy celo transactions since in order to return them over the rpc api since we will need to derive the senders.

Note we probably don't want to merge this right now as it will break a lot of tests since the RPC api will not work.

---
Fixes https://github.com/celo-org/celo-blockchain-planning/issues/256

piersy committed:
Ensure TerminalTotalDifficultyPassed is set

This indicates that the network we are connecting to has at its head
passed the terminal total difficulty.

op-geth won't start unless this is set, so even though our chain has not
passed the terminal total difficulty right now, at the point that we
release CEL2 terminal total difficulty will need to be passed.

We don't actually have difficulty in celo so we'll need to use something
else, the block number seems like a reasonable choice.

piersy committed:
Add custom decoding for celo headers

This allows op geth to decode old celo headers.

piersy committed:
Add support for celo legacy transactions

piersy committed:
Move tx marshalling to own file as much as possible

piersy committed:
Remove feeCurrency notion from op code

piersy committed:
Segregate celo transaction types

piersy committed:
Seggregate celoDecodeTyped

piersy committed:
Update header decoding to introspect the rlp data

This presents a more efficient approach to determining if an encoded
header is pre or post gingerbread.

piersy committed:
Add encoding for pre-gingerbread headers

piersy committed:
Make sure we handle op-geth fields in enc/dec

piersy committed:
Move extra celo code to celo_block.go

piersy committed:
Remove rlpgen header marshaling

We now have a custom mechanism for marshaling the header.

piersy committed:
Rename the celo tx type to match celo blockchain

Since we will be supporting decoding and encoding all old celo tx types in op-geth we may as well use the same naming.

piersy committed:
Move file to match the name

piersy committed:
Move tx type to be next to impl

piersy committed:
Move functionality to appropriate file

piersy committed:
Move content to appropriate file

piersy committed:
Add celo dynamic fee tx

This adds the structure and rlp encode/decode functionality but no
signing or json marshaling logic because for syncing it may not be
required.

mcortesi commented on core/blockchain.go:
```diff
@@ -277,6 +277,8 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis
 		return nil, genesisErr
 	}
+	// Hardcode TerminalTotalDifficultyPassed to true.
+	chainConfig.TerminalTotalDifficultyPassed = true
```
what is this?

piersy replied:
One of the first changes required to just get op-geth to start. This should probably be something that your migration script sets. See a more detailed description [here](https://www.notion.so/clabsco/Migration-celo-optimism-3971e0eabfd64c1f8654d83a8bbed85d?pvs=4#9d09eb67ed984e48998d53f32f290554)

mcortesi replied:
ok, we should move this to the script then

mcortesi commented on eth/backend.go:
```diff
@@ -151,11 +151,14 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
 		return nil, err
 	}
+	// Hardcode TerminalTotalDifficultyPassed to true.
+	chainConfig.TerminalTotalDifficultyPassed = true
```
again, this looks surprising and doesn't give the reader any context, explanation

piersy replied:
Yes, again it's the same change. Since this should be set in the migration script we will be removing this code and as such I didn't comment it heavily.

mcortesi commented on rlp/celo_raw.go:
```diff
@@ -0,0 +1,21 @@
+// the kind of the item, the size of the tag, the size of the content, and any error encountered.
+func ReadNext(buf []byte) (remaining []byte, kind Kind, tagsize, contentsize uint64, err error) {
+	remaining = buf
+	kind, tagsize, contentsize, err = ReadKind(remaining)
```
use `readKind` (lowercase)

mcortesi commented on core/types/celo_legacy_tx.go:
```diff
@@ -0,0 +1,162 @@
+	}
+}
+
+func (tx *LegacyTx) setTxFromEthCompatibleRlpList(rlplist ethCompatibleTxRlpList) {
```
how about making it a standalone function, not to polute LegacyTx methods?

mcortesi commented on core/types/celo_legacy_tx.go:
```diff
@@ -0,0 +1,162 @@
+	tx.EthCompatible = true
+}
+
+func (tx *LegacyTx) setTxFromCeloRlpList(rlplist celoTxRlpList) {
```
how about making it a standalone function, not to polute LegacyTx methods?

piersy replied:
Yep, seems a bit neater 👍

mcortesi commented on core/types/celo_dynamic_fee_tx.go:
```diff
@@ -10,19 +24,21 @@ import (
-	Value      *big.Int
-	Data       []byte
-	AccessList AccessList
+const CeloDynamicFeeTxType = 0x7c
```
i'd move this back to the list on transaction.go; Since that's the list of all tx types, clearer to see it there than here

piersy replied:
op-geth have already started putting their types in separate files see `deposit_tx.go` so it's not the list of all the types. Do you still think we should move this back? BTW op-geth are keeping the tx type in the same file as the tx, as we are doing here.

mcortesi replied:
ok, didn't know, then let's follow their lead

mcortesi commented on core/types/celo_block.go:
```diff
@@ -0,0 +1,151 @@
+	Extra       []byte         `json:"extraData"        gencodec:"required"`
+
+	// Used to cache deserialized istanbul extra data
+	extraLock  sync.Mutex
```
all of these seems to be unused

mcortesi commented on rlp/celo_raw.go:
```diff
@@ -0,0 +1,21 @@
+package rlp
+
+// ReadKind exposes readKind so to enable introspection of rlp encoded data.
+func ReadKind(buf []byte) (k Kind, tagsize, contentsize uint64, err error) {
```
found no uses of it

mcortesi committed:
gofmt nits

mcortesi reviewed (commented):
Looks good. Haven't check the logic, but the structure and DIFF with op-geth.

mcortesi reviewed (commented):
Looks good. Haven't check the logic, but the structure and DIFF with op-geth.

piersy committed:
Remove unused function

piersy committed:
Remove unused fields

piersy committed:
Remove methods from LegacyTx

piersy committed:
Remove unused function

piersy committed:
Use difficulty to identitfy pre-gingerbread header

It seems that in the cmd/devp2p/internal/ethtest tests some testdata
blocks do not have baseFee (probably they are old) so this change will
help with tests that make use of old testdata, since ethereum started
with the difficulty field in block headers.

piersy committed:
Add little bit of debug to check for canonical hashes

piersy commented:
Closing as I will be opening some more targeted PRs with this functionality
